### PR TITLE
Custom node debugging

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

Merge recharge project: custom node debugging.

Two context menu items are added to custom node instance:
  * Expand Custom Node: expand custom node instance to a custom node group
  * Expand Custom Node to Group: expand custom node instance to a normal group
untitled

For a custom node group, there are three options available:
  * Update Custom Node: apply the change in the group to custom node definition
  * Restore to Custom Node Instance: remove group and restore to custom node instance
  * Sync with Custom Node: apply the latest change in the custom node definition back to the group
untitled2

The color of geometries that generated from custom node group will be the same as the group's background color:
untitled3

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.

### CODE REVIEW

**Comment by mjkkirschner** [comment]
👍

**Comment by andydandy74** [comment]
Oh wow! This just made my say. Awesome work!!!

**Comment by jnealb** [comment]
++1++

**Comment by kronz** [comment]
Sweeeet!

**Comment by jnealb** [comment]
Hey @ian take a look

**Comment by ikeough** [comment]
@ke-yu This looks awesome!

Why not just add a property, `CustomNodeModel.IsExpanded` and edit the `WorkspaceView` to apply a view conditionally. If the node is expanded, show the `CustomAnnotationModelView`, and if it's not, show our typical `NodeView`. Then there's just a contextual menu selection, "Expand" (or "Collapse").

And I think, at least for the first release, that syncing should not be left up to the user. It should just happen if you edit nodes that are in the `CustomAnnotationModel`, like when you edit nodes that are in a `CustomNodeWorkspaceModel`. I can see a bunch of ways in which things could get out of sync. Also, because the user has expanded a custom node, if feels to me like they'd expect that stuff to behave like a custom node. I'm guessing you've added this option because the nodes in the `CustomAnnotationModel` are not "in" the custom node workspace, they're in the home workspace, and this is required to sync the function definition. But I still think this should be magic :)

File: src/DynamoCoreWpf/Views/Core/CustomNodeAnnotationView.xaml [hunk 1]
```diff
@@ -0,0 +1,323 @@
+﻿<UserControl Name="Group"
```
Comment (ikeough) [inline src/DynamoCoreWpf/Views/Core/CustomNodeAnnotationView.xaml]: Do we need a completely new xaml here or can we refactor the group view into components that can be used between two different annotation type views?

File: src/DynamoCoreWpf/Views/Core/CustomNodeAnnotationView.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+using Dynamo.Selection;
+using Dynamo.UI.Prompts;
+using TextBox = System.Windows.Controls.TextBox;
+
+namespace Dynamo.Nodes
+{
+    /// <summary>
+    /// Interaction logic for AnnotationView.xaml
+    /// </summary>
+    public partial class CustomNodeAnnotationView : IViewModelView<CustomNodeAnnotationViewModel>
```
Comment (ikeough) [inline src/DynamoCoreWpf/Views/Core/CustomNodeAnnotationView.xaml.cs]: This doesn't inherit anything from `AnnotationView`?

File: src/DynamoCore/Graph/Annotations/CustomNodeAnnotationModel.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        protected override void SerializeCore(XmlElement element, SaveContext context)
+        {
+            XmlElementHelper helper = new XmlElementHelper(element);
+            helper.SetAttribute("functionID", this.FunctionID);
+            base.SerializeCore(element, context);
+        }
+
+        protected override void DeserializeCore(XmlElement element, SaveContext context)
```
Comment (ikeough) [inline src/DynamoCore/Graph/Annotations/CustomNodeAnnotationModel.cs]: Do we need to deserialize the `functionId`?

**Comment by ikeough** [comment]
@ke-yu BTW, the way that Peter and I discussed implementing this previously, was to to put a `WorkspaceView` _on_ the `NodeView` and set its data context to the custom node :) Then make the `NodeView` resizable. This way you wouldn't have to do any of the expand/copy stuff. The view would just know what to draw. This would also be aligned with how EB orginally designed custom node visualization in the "home" workspace.

File: src/DynamoCoreWpf/Views/Core/CustomNodeAnnotationView.xaml.cs (hunk 2 above)
Comment (ke-yu) [inline src/DynamoCoreWpf/Views/Core/CustomNodeAnnotationView.xaml.cs]: hacking, hacking... :-)

**Comment by ke-yu** [comment]
@ikeough , Embedding `WorkspaceView` into `NodeView` is a good idea! But setting context to the custom node always means modifying nodes in this debugging instance will trigger execution for all other custom node instance, right? Anyway, a lot of details haven't been confirmed yet, especially about the interaction between custom node workspace and instance (or annotation in my implementation). Need more inputs from everyone. :-)

**Pull request merged**